app/routers/sessions.py

```diff
@@ -99,45 +109,43 @@ async def create_session(session: Session):
 @router.patch("/{session_id}", response_model=UpdateSessionResponse)
 async def update_session(session_id: str, session_updates: UpdateSession):
     """
-    session is updated whenever start/resume button is clicked
-    (or) end button is clicked for the first time
+    session is updated whenever
+    * start button is clicked (start-quiz event)
+    * resume button is clicked (resume-quiz event)
+    * end button is clicked (end-quiz event)
     """
-    session_updates = jsonable_encoder(session_updates)
+    new_event = jsonable_encoder(session_updates)["event"]
     session = client.quiz.sessions.find_one({"_id": session_id})
     if session is None:
         raise HTTPException(
             status_code=status.HTTP_404_NOT_FOUND,
             detail=f"session {session_id} not found",
         )
 
-    current_time = datetime.utcnow()
-    time_elapsed = 0
-
-    if session_updates["has_quiz_started_first_time"]:
-        # start/resume button clicked first time for a quiz
-        session["has_quiz_started"] = True
-        # time elapsed is zero
+    event_obj = jsonable_encoder(Event.parse_obj({"event_type": new_event}))
+    if session["events"] is None:
+        session["events"] = [event_obj]
     else:
-        # update time remaining based on current time (resume/end button clicked)
+        session["events"].append(event_obj)
+
+    # diff between times of last two events
+    time_elapsed = 0
+    if new_event != EventType.start_quiz:
+        # time_elapsed = 0 for start-quiz event
         time_elapsed = (
-            current_time - datetime.fromisoformat(session["quiz_start_resume_time"])
+            str_to_datetime(session["events"][-1]["created_at"])
+            - str_to_datetime(session["events"][-2]["created_at"])
```

Comment:
Just also add a sanity check here that the `session["events"]` array is of length of 2 or more.
I know that logically it will always be that way but still if something weird happens, I don't want the app to break